heartbeat/service/synthetics_service.go

```diff
@@ -0,0 +1,277 @@
+
+	serviceLocations := serviceManifest.Locations
+
+	pushInterval := 30 * time.Second
+	service.servicePushTicker = time.NewTicker(pushInterval)
+
+	output := Output{}
+	err := b.Config.Output.Config().Unpack(&output)
+	if err != nil {
+		logp.Info("Unable to parse output param")
+		return err
+	}
+
+	// first we need to push at start, and then ticker will take over
+	for locationKey, serviceLocation := range serviceLocations {
+		service.servicePushWait.Add(1)
+		go service.pushConfigsToSyntheticsService(locationKey, serviceLocation, output)
+	}
+	go service.schedulePushConfig(serviceLocations, output)
+	if service.config.ConfigMonitors.Enabled(){
+		go service.scheduleReloadPushConfig(serviceLocations, output)
+	}
+	return nil
+
+}
+
+func (service *SyntheticService) Wait() {
+	service.servicePushWait.Wait()
+}
+
+func (service *SyntheticService) Stop() {
+	service.servicePushTicker.Stop()
+}
+
+func (service *SyntheticService) schedulePushConfig(serviceLocations map[string]config.ServiceLocation, output Output) {
+	service.servicePushWait.Add(1)
+	for {
+		select {
+		case <-service.servicePushTicker.C:
+			for locationKey, serviceLocation := range serviceLocations {
+				service.servicePushWait.Add(1)
+				// first we need to do at start, and then ticker will take over
+				go service.pushConfigsToSyntheticsService(locationKey, serviceLocation, output)
+			}
+
+			defer service.servicePushWait.Done()
+		}
+	}
+}
+
+func (service *SyntheticService) scheduleReloadPushConfig(serviceLocations map[string]config.ServiceLocation, output Output) {
+	reloadPushTicker := time.NewTicker(1 * time.Hour)
+	for {
+		select {
+		case <-service.serviceRunnerFactory.Update:
+			if reloadPushTicker == nil {
+				reloadPushTicker = time.NewTicker(1 * time.Second)
+			} else {
+				reloadPushTicker.Reset(1 * time.Second)
+			}
+
+		case <-reloadPushTicker.C:
+			for locationKey, serviceLocation := range serviceLocations {
+				service.servicePushWait.Add(1)
+				// first we need to do at start, and then ticker will take over
+				go service.pushConfigsToSyntheticsService(locationKey, serviceLocation, output)
+			}
+			reloadPushTicker.Stop()
+		}
+	}
+}
+
+func (service *SyntheticService) getSyntheticServiceManifest() (config.ServiceManifest, error) {
+	serviceCfg := service.config.Service
+	var err error
+
+	if serviceCfg.Username == "" {
+		err = errors.New("synthetic service username is required for authentication")
+	}
+
+	if serviceCfg.Password == "" {
+		err = errors.New("synthetic service password is required for authentication")
+	}
+
+	if serviceCfg.ManifestURL == "" {
+		err = errors.New("synthetic service manifest url is required")
+	}
+
+	if err != nil {
+		return config.ServiceManifest{}, err
+	}
+
+	req, err := http.NewRequest("GET", serviceCfg.ManifestURL, nil)
+
+	resp, err := Client.Do(req)
+
+	if err != nil {
+		return config.ServiceManifest{}, err
+	}
+
+	serviceManifest := config.ServiceManifest{}
+
+	read, err := ioutil.ReadAll(resp.Body)
```

Comment:
Since we have an `io.Reader` it's cleaner to use the `json.NewDecoder` to read from that rather than slurping it into a byte array and using `json.Unmarshal`. See https://stackoverflow.com/a/31129967